Use util.url.join for URLs in GNU Mirrors / reorder Mirrors

ChristianTackeGSI opened the pull request:
One should not use os.path.join for URLs. This does only work on POSIX systems.

Instead use urljoin from urllib.parse.
Update used prefixes accordingly.

adamjstewart commented:
@alalazo 

Also, shouldn't the mirror URL come first?

alalazo commented:
> Also, shouldn't the mirror URL come first?

It depends on which one we want to try first, I have no objection if we want to switch (both urls have been "unreliable" at different times).

ChristianTackeGSI commented:
> LGTM, only one minor comment. Checking the codebase I noticed that we also have a `spack.util.url.join` function that was merged with #11117, so maybe we should check with @opadron if it is the case to use that instead of `urljoin`.

Looking at the docs for `spack.util.url.join` it seems like it should be identical to `urllib.parse.urljoin` when called with `resolve_href == True`. That said, I'd vote for using a stdlib function when it is available and does the job.
The current case is so simple, we could even just use string concatenation.

opadron commented:
Our `join()` differs from `urllib` in that it is aware of more URL schemes.  I think all URL joins should prefer it so that joins and link resolution behavior is consistent across all of Spack.

We're still wrestling with the challenge of what to do about `spack.util.url.join`, since we've already had others inquire about extending the set of supported URL schemes (which the current implementation was not designed to handle) and we're still not very pleased with how complex the code is.  I'm working on another approach that should address these issues and hopefully bring the code closer to the "simple string concatenation" simplicity ideal.

ChristianTackeGSI commented:
> Our `join()` differs from `urllib` in that it is aware of more URL schemes. I think all URL joins should prefer it so that joins and link resolution behavior is consistent across all of Spack.

So should I rewrite my patch to use `spack.util.url.join`, string concatenation, or leave it as is?

opadron commented:
Yes (sorry if I wasn't clear), I think it would be better to use `spack.util.url.join`.

Be sure to set `resolve_href=True` and mind the presence or lack of a trailing forward slash `/` for the base mirror part.  You should be getting the exact same results as `urllib`.  Let us know if you don't or have other questions.

adamjstewart commented:
> > Also, shouldn't the mirror URL come first?
> 
> It depends on which one we want to try first, I have no objection if we want to switch (both urls have been "unreliable" at different times).

According to https://www.gnu.org/prep/ftp.en.html:

> GNU software is available on ftp.gnu.org under the directory /gnu. Please try to use one of the many mirrors of our site listed below: the mirrors will give you faster response. Some mirrors also provide a copy of the rest of ftp.gnu.org; try the parent directory if you're interested. You can use the generic url http://ftpmirror.gnu.org to automatically choose a nearby and up-to-date mirror.

So I think we should move the mirror URL first.

ChristianTackeGSI commented:
Changed and rebased.

I have put the mirror reordering in its own commit. Please don't squash them together, because these are really two different things.

ChristianTackeGSI commented:
This has been approved by @adamjstewart 12 days ago.
Do I need to do anything to get this PR merged?

N.B.: I have put the mirror reordering in its own commit. Please don't squash them together, because these are really two different things.

alalazo commented on lib/spack/spack/build_systems/gnu.py:
```diff
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
-import os.path
+import spack.util.url as url_util
```
```suggestion
import spack.util.url
```

alalazo commented on lib/spack/spack/build_systems/gnu.py:
```diff
@@ -26,7 +26,8 @@ class GNUMirrorPackage(spack.package.PackageBase):
         self._ensure_gnu_mirror_path_is_set_or_raise()
         return [
-            os.path.join(m, self.gnu_mirror_path) for m in self.base_mirrors
+            url_util.join(m, self.gnu_mirror_path, resolve_href=True)
```
```suggestion
            spack.util.url.join(m, self.gnu_mirror_path, resolve_href=True)
```

ChristianTackeGSI commented:
@alalazo : included and rebased.
